File: leetcode/272. Closest Binary Search Tree Value II.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def closestKValues(self, root: Optional[TreeNode], target: float, k: int) -> List[int]:
        lst = []
        def inorder(node):
            if node is None:
                return
            inorder(node.left)
            lst.append(node.val)
            inorder(node.right)
            
        inorder(root)
        
        left = 0
        right = len(lst) - 1
        
        while right - left >= k:
            if abs(lst[left] - target) > abs(lst[right] - target):
                left += 1
            else:
                right -= 1
        
        return lst[left:right+1]
      
      
      
 

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from heapq import heappush, heappop

class Solution:
    def closestKValues(self, root: Optional[TreeNode], target: float, k: int) -> List[int]:
        
        heap = []
        def inorder(node):
            if node is None:
                return
            inorder(node.left)
            
            heappush(heap, (-abs(node.val - target), node.val))
            if len(heap) > k:
                heappop(heap)
            
            inorder(node.right)
            
        inorder(root)
        
        return [val for diff, val in heap]

File: leetcode/272. Closest Binary Search Tree Value II.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn closest_k_values(root: Option<Rc<RefCell<TreeNode>>>, target: f64, k: i32) -> Vec<i32> {
        fn inorder(root: &Option<Rc<RefCell<TreeNode>>>, nums: &mut Vec<i32>) {
            match root {
                Some(node) => {
                    let node = node.borrow();
                    inorder(&node.left, nums);
                    nums.push(node.val);
                    inorder(&node.right, nums);
                },
                None => {
                }
            }
        }
        
        let mut nums = Vec::new();
        inorder(&root, &mut nums);
        
        let mut i = 0;
        let mut j = nums.len() - 1;
        while j - i >= k as usize {
            if f64::abs(nums[j] as f64 - target) < f64::abs(nums[i] as f64 - target) {
                i += 1;
            } else {
                j -= 1;
            }
        }
        
        nums[i..=j].to_vec()
    }
}
